File: src/main/dart/lib/swagger.dart
/// Swagger
library swagger;

/// [Listing]
class Listing {
  String apiVersion, basePath, resourcePath;
  List<Resource> resources;

  Listing();

  factory Listing.fromJSON(Map json) =>
      new Listing()
      ..apiVersion = json["apiVersion"]
      ..basePath = json["basePath"]
      ..resourcePath = json["resourcePath"]
      ..resources = json["apis"].map((j) => new Resource.fromJSON(j)).toList();
}

/// [Resource]
class Resource {

  String path,
         description;
  List<Operation> operations;

  Resource();

  factory Resource.fromJSON(Map json) =>
    new Resource()
    ..path = json["path"]
    ..description = json["description"]
    ..operations = json["operations"].map((j) => new Operation.fromJSON(j)).toList();
}

/// [Operation]
class Operation {
  String method,
         name,
         type,
         summary,
         notes;
  List<Parameter> parameters;

  Operation._();

  factory Operation.fromJSON(Map json) =>
      new Operation._()
      ..method = json["method"]
      ..name = json["nickname"]
      ..type = json["type"]
      ..summary = json["summary"]
      ..notes = json["notes"]
      ..parameters = json["parameters"].map((j) => new Parameter.fromJSON(j)).toList();
}

/// [Parameter]
class Parameter {

  String name,
         type,
         description,
         dataType;
  bool required;

  Parameter._();

  factory Parameter.fromJSON(Map json) =>
      new Parameter._()
      ..name = json["name"]
      ..type = json["paramType"]
      ..dataType = json["dataType"]
      ..description = json["description"]
      ..required = json["required"] == "true";

  bool get isPathParam => type == "path";
}

File: src/main/js/src/swagger.ts
/**
@license
(C) Copyright Nuxeo Corp. (http://nuxeo.com/)
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/// The API Object describes one or more operations on a single path.
interface Resource {
  path: string;
  description: string;
  operations: Operation[];
}

/// The Operation object describes a single operation on a path.
interface Operation {
  method: string;
  name: string;
  type: string;
  summary: string;
  notes: string;
  parameters: Parameter[];
}

/// The Parameter Object describes a single parameter to be sent in an operation.
interface Parameter {
  name: string;
  paramType: string;
  description: string;
  dataType: string;
  required: string;
}

export { Resource, Operation, Parameter}
